File: src/TerminalForms/Rect.cs
namespace TerminalForms;

public sealed class Rect : IDisposable, IEquatable<Rect>
{
    public bool IsDisposed { get; private set; }

    public IntPtr Ptr { get; }

    public Rect()
    {
        Ptr = NativeMethods.Tv_Rect_new0();
    }

    public Rect(int ax, int ay, int bx, int by)
    {
        Ptr = NativeMethods.Tv_Rect_new1(ax, ay, bx, by);
    }

    public Rect(Point p1, Point p2)
    {
        Ptr = NativeMethods.Tv_Rect_new2(p1.Ptr, p2.Ptr);
    }

    public Rect(IntPtr ptr)
    {
        Ptr = ptr;
    }

    ~Rect()
    {
        Dispose(false);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    private void Dispose(bool disposing)
    {
        if (!IsDisposed)
        {
            NativeMethods.Tv_Rect_delete(Ptr);
            IsDisposed = true;
        }
    }

    public void Move(int aDX, int aDY)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        NativeMethods.Tv_Rect_move(Ptr, aDX, aDY);
    }

    public void Grow(int aDX, int aDY)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        NativeMethods.Tv_Rect_grow(Ptr, aDX, aDY);
    }

    public void Intersect(Rect r)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        ObjectDisposedException.ThrowIf(r.IsDisposed, r);
        NativeMethods.Tv_Rect_intersect(Ptr, r.Ptr);
    }

    public void Union(Rect r)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        ObjectDisposedException.ThrowIf(r.IsDisposed, r);
        NativeMethods.Tv_Rect_Union(Ptr, r.Ptr);
    }

    public bool Contains(Point p)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        ObjectDisposedException.ThrowIf(p.IsDisposed, p);
        return NativeMethods.Tv_Rect_contains(Ptr, p.Ptr);
    }

    public bool IsEmpty()
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        return NativeMethods.Tv_Rect_isEmpty(Ptr);
    }

    public override bool Equals(object? obj)
    {
        return Equals(other: obj as Rect);
    }

    public override int GetHashCode()
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        return NativeMethods.Tv_Rect_hash(Ptr);
    }

    public bool Equals(Rect? other)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        if (other is null)
            return false;
        ObjectDisposedException.ThrowIf(other.IsDisposed, other);
        return NativeMethods.Tv_Rect_operator_equals(Ptr, other.Ptr);
    }

    public Point A
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            return new Point(NativeMethods.Tv_Rect_get_a(Ptr));
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            ObjectDisposedException.ThrowIf(value.IsDisposed, value);
            NativeMethods.Tv_Rect_set_a(Ptr, value.Ptr);
        }
    }

    public Point B
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            return new Point(NativeMethods.Tv_Rect_get_b(Ptr));
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            ObjectDisposedException.ThrowIf(value.IsDisposed, value);
            NativeMethods.Tv_Rect_set_b(Ptr, value.Ptr);
        }
    }
}

File: src/tvision4c/Rect.cpp
#include "Rect.h"

EXPORT TRect* Tv_Rect_new0() {
    return new TRect;
}

EXPORT TRect* Tv_Rect_new1(int32_t ax, int32_t ay, int32_t bx, int32_t by) {
    return new TRect(ax, ay, bx, by);
}

EXPORT TRect* Tv_Rect_new2(TPoint* p1, TPoint* p2) {
    return new TRect(*p1, *p2);
}

EXPORT void Tv_Rect_delete(TRect* self) {
    delete self;
}

EXPORT int32_t Tv_Rect_hash(TRect* self) {
    return hash(Tv_Point_hash(&self->a), Tv_Point_hash(&self->b));
}

EXPORT void Tv_Rect_move(TRect* self, int32_t aDX, int32_t aDY) {
    self->move(aDX, aDY);
}

EXPORT void Tv_Rect_grow(TRect* self, int32_t aDX, int32_t aDY) {
    self->grow(aDX, aDY);
}

EXPORT void Tv_Rect_intersect(TRect* self, TRect* r) {
    self->intersect(*r);
}

EXPORT void Tv_Rect_Union(TRect* self, TRect* r) {
    self->Union(*r);
}

EXPORT BOOL Tv_Rect_contains(TRect* self, TPoint* p) {
    return self->contains(*p);
}

EXPORT BOOL Tv_Rect_operator_equals(TRect* self, TRect* r) {
    return *self == *r;
}

EXPORT BOOL Tv_Rect_isEmpty(TRect* self) {
    return self->isEmpty();
}

EXPORT TPoint* Tv_Rect_get_a(TRect* self) {
    return new TPoint{ self->a };
}

EXPORT TPoint* Tv_Rect_get_b(TRect* self) {
    return new TPoint{ self->b };
}

EXPORT void Tv_Rect_set_a(TRect* self, TPoint* p) {
    self->a = *p;
}

EXPORT void Tv_Rect_set_b(TRect* self, TPoint* p) {
    self->b = *p;
}
